protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnConverter.java

```diff
@@ -35,8 +36,16 @@
 
     @VisibleForTesting final ArrayList<MessageUid> uids;
 
-    public UidMsnConverter(Iterator<MessageUid> iterator) {
-        uids = Lists.newArrayList(iterator);
+    public UidMsnConverter() {
+        this.uids = Lists.newArrayList();
+    }
+
+    public synchronized void addAll(List<MessageUid> iterator) {
+        LinkedHashSet<MessageUid> messageUids = new LinkedHashSet<MessageUid>();
```

Comment:
Each addAll on a TreeSet is n*ln(n), no? So we end up not gaining much / at all, at least from a complexity point of view.

Variable reallocation VS clear: we have to choose at least one evil...